src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -397,27 +364,208 @@ CollectionExpressionTypeKind.Span or
         }
 
-        private BoundExpression MakeCollectionExpressionSpreadElement(BoundCollectionExpressionSpreadElement initializer)
+        /// <summary>
+        /// Create and populate an array from a collection expression where the
+        /// collection has a known length, although possibly including spreads.
+        /// </summary>
+        private BoundExpression CreateAndPopulateArray(BoundCollectionExpression node, ArrayTypeSymbol arrayType)
         {
-            var enumeratorInfo = initializer.EnumeratorInfoOpt;
-            var addElementPlaceholder = initializer.AddElementPlaceholder;
+            var syntax = node.Syntax;
+            var elements = node.Elements;
 
-            Debug.Assert(enumeratorInfo is { });
-            Debug.Assert(addElementPlaceholder is { });
-            Debug.Assert(addElementPlaceholder.Type is { });
+            if (node.GetKnownLength(out bool hasSpreadElements) is not int knownLength)
+            {
+                // Should have been handled by the caller.
+                throw ExceptionUtilities.UnexpectedValue(node);
+            }
 
-            var syntax = (CSharpSyntaxNode)initializer.Syntax;
-            var iterationVariable = _factory.SynthesizedLocal(addElementPlaceholder.Type, syntax);
-            var convertedExpression = (BoundConversion)initializer.Expression;
+            if (knownLength == 0)
+            {
+                return CreateEmptyArray(syntax, arrayType);
+            }
 
-            AddPlaceholderReplacement(addElementPlaceholder, _factory.Local(iterationVariable));
+            if (!hasSpreadElements)
+            {
+                var initialization = new BoundArrayInitialization(
+                        syntax,
+                        isInferred: false,
+                        elements.SelectAsArray(static (element, rewriter) => rewriter.VisitExpression(element), this));
+                return new BoundArrayCreation(
+                    syntax,
+                    ImmutableArray.Create<BoundExpression>(
+                        new BoundLiteral(
+                            syntax,
+                            ConstantValue.Create(knownLength),
+                            _compilation.GetSpecialType(SpecialType.System_Int32))),
+                    initialization,
+                    arrayType)
+                { WasCompilerGenerated = true };
+            }
 
-            var rewrittenBody = VisitStatement(initializer.AddMethodInvocation);
-            Debug.Assert(rewrittenBody is { });
+            NamedTypeSymbol intType = _compilation.GetSpecialType(SpecialType.System_Int32);
+            BoundAssignmentOperator assignmentToTemp;
+            var localsBuilder = ArrayBuilder<BoundLocal>.GetInstance();
+            var sideEffects = ArrayBuilder<BoundExpression>.GetInstance();
 
-            RemovePlaceholderReplacement(addElementPlaceholder);
+            // Rewrite expressions into temporaries.
+            foreach (var element in elements)
```

Comment:
Not sure how I feel about this. Given we are creating an array I am not sure it is fine to create potentially very large numbers of temporaries just because a spread element (with runtime-known length) is present.

Reply:
Basically, for this release, I would be absolutely fine with having an intermediate List for these cases. I could easily see a code generator wanting to write something like

```cs
public static readonly string[] arr1 = ["bunch", "of", "values", "like", "thousands"];
public static readonly string[] arr2 = [..arr1, "even", "more", "values"];
```

In this specific case, maybe blowing the stack on a static constructor is less likely, but still maybe possible.

Also, I think in some cases compiler can probably determine that a spread operator like `..arr1` is non-side-effecting to evaluate, e.g. we can read `Length` before evaluating the elements in-order without a problem, and optimize accordingly. I would def feel comfortable saying that if your Length property itself is side effecting, then bad things may happen, sorry.

Reply:
Updated to only create temporaries up to the last spread element, and up to a maximum of 4 elements. That will cover cases such as `[..s, e1, e2, ...]` and `[..s1, ..s2]` in particular.